frontend/packages/git-service/src/services/base-service.ts

```diff
@@ -37,12 +37,19 @@ export abstract class BaseService {
   // Checks if dockerfile exist in the repo and returns dockerfile content
   abstract async getDockerfileContent(): Promise<string>;
 
+  abstract async getPackageJsonContent(): Promise<string>;
+
   // Detect build types for given gitsource, It runs regular expressions on file list
   // and returns list of build types matched.
   async detectBuildType(): Promise<BuildType[]> {
     try {
       const fileList = await this.getRepoFileList();
-      return detectBuildType(fileList.files);
+      let isWebApp = false;
+      if (fileList.files.includes('package.json')) {
+        const packageJsonContent = await this.getPackageJsonContent();
+        isWebApp = isModernWebApp(packageJsonContent);
+      }
+      return detectBuildType(fileList.files, isWebApp);
```

Comment:
How about:
```
const buildTypes = detectBuildType(fileList.files);
if (fileList.files.includes('package.json')) {
  const packageJsonContent = await this.getPackageJsonContent();
  if (isModernWebApp(packageJsonContent)) {
    return buildTypes.sort((a, b) => {
      if (b.buildType === 'modern-webapp') return 1;
      if (a.buildType === 'modern-webapp') return -1;
      return 0;
    });
  }
}
return buildTypes;
```
That way we keep `detectBuildType` pure and leave ourselves open to move this logic elsewhere if we need to address other lib discrepancies.

Obviously you'd need to clean up `detectBuildType`.

Reply:
Yeah, that makes sense. Moved the logic of sorting based on modern web app to base service method.